### DISCUSSION TITLE
Analyze and reduce package size when loaded in memory

### DISCUSSION THREAD

**Discussion by adriendelsalle** [pyodide/pyodide#4933] [Q&A]
Description
---

I'm trying to build a very large package, namely `occt` (Open CASCADE, an Open Source CAD library), with its Python bindings using SWIG https://github.com/tpaviot/pythonocc-core.

I'm not posting a package request because such a large package doesn't looks like a good candidate for a "in-tree" package.
The question may also be more related to `wasm`/`emscripten` but I'm not sure. Depending on advice I'll move it upstream.

The resulting package creates a `SIGILL` error when importing it in a `pyodide` console in Chrome, which looks like an out-of-memory.
When I'm disabling large portions of the libs and associated bindings, it works fine but I see a huge memory consumption.

Some figures to better understand the situation:
- partial package, including shared libs of `occt` vendored using `build: vendor-sharedlib: true`: 168MB unpacked
  - I removed bindings but kept `occt` not limited to binded parts
  - AFAIU, `pyodide` loads all the wasm files found in a package before any import, even the shared libs that are not linked by any others. It's what I understand from https://github.com/pyodide/pyodide/blob/fffbfa6acc8d6428fd3ce96c237dc5fd8ad9fc5f/src/js/dynload.ts#L76-L88 
- heap alloc: 10MB just after loading the console, 700MB (I looked chrome memory in dev panel) after importing `OCC.Core` (the top-level import)
- chrome tab memory seen by system monitor: 1.3GB

As a reference, I loaded the "core" submodules corresponding of the partial package I built in wasm in a `ipython` console:
- process seen by system monitor: 109MB
- `ipython` takes about 43MB
- 30MB for `OCC.Core` (`pythonocc-core` bindings) loaded files only + 76MB for `occt` (all libs, but not all are linked by the partial package)

All of this just looks like fairly consistent.

I saw similar behavior in `scipy` or `numpy` which take 47MB and 12MB resp unpacked and ~300MB and ~100MB when loaded (https://github.com/pyodide/pyodide/discussions/4725).

Question
---

What is the ratio and rationale of the memory consumption between binary files on disk and in-memory?
Why such a huge memory consumption? And what is the diff between heap alloc and system monitor?

Is there any compile time flags or other things to consider for reducing that?
Any methodology to investigate what is the main source of such a delta?

Additional info
---

I'm running tests on:
- Linux 64, Ubuntu 22.04.4
- Chrome Version 125.0.6422.141 (Official Build) (64-bit)

If it helps, with some extra clean-up I can share a branch to reproduce the build.

I'm new to `wasm` and `pyodide`, with no background in `js`, etc. My area is C++ and Python related.

Thanks for your help!

**Comment by adriendelsalle** [comment]
@ryanking13 gentle ping, since you answered https://github.com/pyodide/pyodide/discussions/4725 you're probably the best to tackle this one too

**Comment by ryanking13** [comment]
> The resulting package creates a SIGILL error when importing it in a pyodide console in Chrome, which looks like an out-of-memory.

I never encountered the `SIGILL` error, but it seems reasonable to assume that it is caused by out-of-memory.

> AFAIU, pyodide loads all the wasm files found in a package before any import, even the shared libs that are not linked by any others. It's what I understand from

That's true.

> What is the ratio and rationale of the memory consumption between binary files on disk and in-memory?

We don't have the stats, but when you load a library from Pyodide, it first has to store files in the in-memory file system. Then it compiles the libraries into WASM modules, and then there's a bunch of data that Emscripten stores along the way, so I think it uses quite a bit of memory.

> And what is the diff between heap alloc and system monitor?

Sorry, I have no idea about that.

> Is there any compile time flags or other things to consider for reducing that?

Optimization flags like `-Os`, `-O3`, `-O2` affects the final output size. I am not sure how much it will affect the memory usage though. But I think it is worth to investigate.

**Reply by adriendelsalle** [reply]
Thanks a lot @ryanking13 !

> it first has to store files in the in-memory file system. Then it compiles the libraries into WASM modules

You mean that for a lib of 100MB unpacked, it will always take 100MB of virtual fs and probably the same to load it in memory?
Is there any interest of keeping the file itself except for caching from one session to another? Is it kept packed or unpacked?
It would probably worth considering multiple strategies, including deletion of the unpacked wasm libs after loading?

**Reply by ryanking13** [reply]
> You mean that for a lib of 100MB unpacked, it will always take 100MB of virtual fs and probably the same to load it in memory?

Yes, storing the file will consume memory equal to the file size, and if you imported the file, it will take additional memory for the Python interpreter to read and process the file.

> Is there any interest of keeping the file itself except for caching from one session to another?

There was some discussion of keeping the file in the IndexedDB #3664.

> It would probably worth considering multiple strategies, including deletion of the unpacked wasm libs after loading?

In a native environment, it would be a very strange way to delete a file after importing a module, but in a browser environment, which is disposable anyway, it seems like a possible option. However, memory usage is not currently a priority for me with Pyodide, and I don't think it should be built into Pyodide itself. So, if you're interested, I'd love to see you create your own tool (maybe in the form of a Python package?) and share it with the Pyodide community.

**Comment by ryanking13** [comment]
> Any methodology to investigate what is the main source of such a delta?

Sorry, no. However, my general rule of thumb is that memory usage and the on-disk size of the library are proportional, so I think it's best to look for ways to reduce the file size.
